scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java

```diff
@@ -1078,183 +1085,134 @@ private void processChangeCipherSpecRecord(Record record) {
-		if (connection.hasOngoingHandshake()) {
-			DTLSSession handshakeSession = connection.getOngoingHandshake().getSession();
-			if (handshakeSession.getReadEpoch() == record.getEpoch()) {
-				// evaluate message in context of ongoing handshake
-				record.setSession(handshakeSession);
-			} else if (!record.isNewClientHello()) {
-				// epoch is not the same as the current session so we
-				// can not decrypt the message now. Let handshaker handle it
-				// (it can queue it to deal with it later)
-				connection.getOngoingHandshake().processMessage(record);
-				return;
-			}
-		} else if (connection.hasEstablishedSession() && connection.getEstablishedSession().getReadEpoch() == record.getEpoch()) {
-			// client wants to re-negotiate established connection's crypto params
-			// evaluate message in context of established session
-			record.setSession(connection.getEstablishedSession());
-		} else if (record.isNewClientHello()) {
-			// client has lost track of existing connection and wants to negotiate a new connection
-			// in epoch 0 no crypto params have been established yet, thus we do not need to set a session
-		} else {
-			LOGGER.debug(
-				"Discarding HANDSHAKE message [epoch={}] from peer [{}] which does not match expected epoch(s)",
-				new Object[]{record.getEpoch(), record.getPeerAddress()});
-			return;
-		}
-
-		try {
-			HandshakeMessage handshakeMessage = (HandshakeMessage) record.getFragment();
-			processDecryptedHandshakeMessage(handshakeMessage, record, connection);
-		} catch (GeneralSecurityException e) {
-			discardRecord(record, e);
-		}
-	}
-
 	private void processDecryptedHandshakeMessage(final HandshakeMessage handshakeMessage, final Record record,
 			final Connection connection) throws HandshakeException {
 		switch (handshakeMessage.getMessageType()) {
 		case CLIENT_HELLO:
 			processClientHello((ClientHello) handshakeMessage, record, connection);
 			break;
 		case HELLO_REQUEST:
-			processHelloRequest((HelloRequest) handshakeMessage, connection);
+			processHelloRequest(connection);
 			break;
 		default:
 			processOngoingHandshakeMessage(handshakeMessage, record, connection);
 		}
 	}
 
 	/**
+	 * Process handshake message of an ongoing handshake.
 	 * 
-	 * @param message
-	 * @param record
-	 * @param connection
+	 * @param message handshake message
+	 * @param record record of handshake message
+	 * @param connection connection to process handshake message. May only be
+	 *            {@code null}, if a CLIENT_HELLO is received.
 	 * @throws HandshakeException if the handshake message cannot be processed
 	 */
-	private static void processOngoingHandshakeMessage(final HandshakeMessage message, final Record record, final Connection connection) throws HandshakeException {
+	private void processOngoingHandshakeMessage(final HandshakeMessage message, final Record record, final Connection connection) throws HandshakeException {
```

Comment:
javadoc say connection can be NULL, if this is true => NPE spotted

Reply:
No, javadoc says:

> May only be {@code null}, if a CLIENT_HELLO is received.

The null connection is catched up in `receiveNextDatagramFromNetwork` with

```
if (connection == null && (records.size() != 1 || !records.get(0).isNewClientHello())) {
	LOGGER.debug("Discarding {} records from [{}] received without existing connection", records.size(), peerAddress);
	return;
}
```

(And I plan, to extract the processing of the "connection less CLIENT_HELLO" from the serialized execution, but that will be a future step ... )

Reply:
> No, javadoc says:
>
>    May only be {@code null}, if a CLIENT_HELLO is received.

I still understand that when I use this method I can use a null connection.

Reply:
And so you propose?
And again, only for a short term, then the processing of "connection less CLIENT_HELLO" will be redesigned in one of the next steps. Then the "maybe null" gets removed.

Reply:
If this will be removed, ignore this comment. Go to merge :D

Reply:
> May only be {@code null}

was used in some places, therefore I checked the wrong ones. At this location you are right and I fixed the javadoc. :-)